### ISSUE TITLE
Be able to customize verbosity information

### ISSUE THREAD

**Issue by sawenzel** [FairRootGroup/FairLogger#6]
At the moment it is possible to have fixed formats for verbosity levels
// verylow:  message
// low:      [severity] message
// medium:   [HH:MM:SS][severity] message
// high:     [process name][HH:MM:SS:µS][severity] message
// veryhigh: [process name][HH:MM:SS:µS][severity][file:line:function] message

As a user, I would like to be able to configure these formats, or to provide a custom verbosity level.
Say I need something without process name, without time but more information on the function (for instance including the full namespace).

**Comment by dennisklein** [comment]
In #7 I propose the following solution:

Next to the predefined verbosity levels, there are 4 new ones, `user1`, `user2`, `user3`, `user4`. With the new API `Logger::DefineVerbosity(...)` you may override either any of the predefined ones or any of the new user levels.

```cpp
Logger::DefineVerbosity(Verbosity::user2,
                        VerbositySpec::Make(VerbositySpec::Info::file_line_function,
                                            VerbositySpec::Info::process_name));
```

The order of the arguments to `VerbositySpec::Make(...)` is important, e.g. the above results in:

```
[loggerTest.cxx:29:printEverySeverity][loggerTest] info message 289
```

What do you think? (So far, this is only the API version at FairLogger level. If this needs to be customizable via the CLI, you/we could of course add the appropriate code in O2, FairMQ or FairRoot. Just let us know.)

Unfortunately, I have not found a solution for your second request to provide a way to print the full qualified name of the function. If you know how to get this information, I am happy to add it.

**Comment by dennisklein** [comment]
The arguments to `VerbositySpec::Make(...)` can be:

```cpp
struct VerbositySpec
{
    enum class Info : int
    {
        __empty__ = 0,         // used to initialize order array
        process_name,          // [process name]
        timestamp_s,           // [HH:MM:SS]
        timestamp_us,          // [HH:MM:SS:µS]
        severity,              // [severity]
        file,                  // [file]
        file_line,             // [file:line]
        file_line_function,    // [file:line:function]
        __max__                // needs to be last in enum
    };
...
};
```

**Comment by fuhlig1** [comment]
Hi Dennis,

BOOST_CURRENT_FUNCTION could be the universal solution to get the full information. See the example below.

Probably the output has to be filtered first because the results also contains the full list of parameters.

Ciao

Florian

> #include <boost/current_function.hpp>
> #include <iostream>
> 
> namespace foo {
> namespace {
> 
> template<int i>
> int (*bar(int (*arg)(int *)))(int *)
> {
>   std::cout << "Inside bar" << std::endl;
>   std::cout << __func__ << std::endl;
>   std::cout << __PRETTY_FUNCTION__ << std::endl;
>   std::cout << BOOST_CURRENT_FUNCTION << std::endl;
>   return arg;
> }
> 
> void foo()
> {
>   std::cout << "Inside foo" << std::endl;
>   std::cout << __func__ << std::endl;
>   std::cout << __PRETTY_FUNCTION__ << std::endl;
>   std::cout << BOOST_CURRENT_FUNCTION << std::endl;
> }
> 
> } // anonymous namespace
> } // namespace foo
> 
> int main()
> {
>   std::cout << __func__ << std::endl;
>   std::cout << BOOST_CURRENT_FUNCTION << std::endl;
> 
>   foo::bar<1337>(nullptr);
>   foo::foo();
> 
> }

**Comment by sawenzel** [comment]
@dennisklein : I think this is going into the right direction and with the suggestion of @fuhlig1 we should be able to get detailed function info.

**Comment by dennisklein** [comment]
I added support for `BOOST_PRETTY_FUNCTION`. Find also the added documentation for [custom verbosities](https://github.com/FairRootGroup/FairLogger/blob/a1b46efbdccc28ee882c812f7f4b82bca38df4cf/README.md#3-verbosity) and [`BOOST_PRETTY_FUNCTION` support](https://github.com/FairRootGroup/FairLogger/blob/a1b46efbdccc28ee882c812f7f4b82bca38df4cf/README.md#31-boost_pretty_function-support).

**Comment by dennisklein** [comment]
@sawenzel If you give us positive feedback, that the current `dev` of FairLogger solves your feature request, I will make a release.

**Comment by dennisklein** [comment]
This was released in [v1.4.0](https://github.com/FairRootGroup/FairLogger/releases/tag/v1.4.0).
